src/EthernetClient.cpp

```diff
@@ -65,17 +65,19 @@ int EthernetClient::connect(IPAddress ip, uint16_t port)
   _tcp_client->data.available = 0;
   _tcp_client->state = TCP_NONE;
 
+  uint32_t startTime = millis();
   ip_addr_t ipaddr;
   tcp_arg(_tcp_client->pcb, _tcp_client);
-  if (ERR_OK != tcp_connect(_tcp_client->pcb, u8_to_ip_addr(rawIPAddress(ip), &ipaddr), port, &tcp_connected_callback)) {
+  if (ERR_OK != tcp_connect(_tcp_client->pcb, u8_to_ip_addr(rawIPAddress(ip), &ipaddr), port, &tcp_connected_callback)
+      || ((millis() - startTime) >= _timeout)) {
```

Comment:
It looks like timeout is useless in this case, because when `tcp_connect()` retruns whatever the time it takes, the important things is the error code.
```suggestion
if (ERR_OK != tcp_connect(_tcp_client->pcb, u8_to_ip_addr(rawIPAddress(ip), &ipaddr), port, &tcp_connected_callback)) {
```